src/fsharp/FSharp.Core/array.fs

```diff
@@ -494,12 +494,23 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Filter")>]
         let filter f (array: _[]) = 
-            checkNonNull "array" array
-            let res = List<_>() // ResizeArray
-            for i = 0 to array.Length - 1 do 
-                let x = array.[i] 
-                if f x then res.Add(x)
-            res.ToArray()
+            checkNonNull "array" array                        
+            let temp = Array.zeroCreateUnchecked array.Length            
```

Comment:
If the main goal here is to reduce allocations, you _could_ consider BitArray (https://msdn.microsoft.com/en-us/library/system.collections.bitarray(v=vs.110).aspx), which would make this array ~1/8th the current memory, but it might come at a slight perf cost.